libraries/chain/controller.cpp

```diff
@@ -5216,25 +5208,39 @@ int64_t controller_impl::set_proposed_producers( vector<producer_authority> prod
 
    assert(pending);
 
-   auto [version, diff] = pending->get_next_proposer_schedule_version(producers);
-   if (!diff)
-      return version;
+   // see if one already proposed in this block
+   auto& gpo = db.get<global_property_object>();
+   if (gpo.proposed_schedule_block_num) {
+      if (std::equal(producers.begin(), producers.end(),
+                     gpo.proposed_schedule.producers.begin(), gpo.proposed_schedule.producers.end())) {
+         return std::numeric_limits<uint32_t>::max(); // the proposed producer schedule does not change
+      }
+      // clear gpo proposed_schedule as we may determine no diff between proposed producers and next proposer schedule
+      db.modify( gpo, [&]( auto& gp ) {
+         gp.proposed_schedule_block_num.reset();
+         gp.proposed_schedule.version = 0;
+         gp.proposed_schedule.producers.clear();
+      });
+   }
```

Comment:
not needed anymore I think, we can just overwrite the last one.

Reply:
We don't overwrite if it reverts back to the current active or previously proposed schedule.

Reply:
I don't follow. I thought we agreed that whener a new policy was proposed in a block, it will have a new generation number, even if the diff is empty. 
Which means that we can always update the `global_property_object` withing the block (`version` filed is irrelevant). 
After executing the transactions, if the `global_property_object` does contain a new policy for this block, we increment the version tracked in `parent` and store the new policy in the if_extension.

Reply:
That was not my understanding. Why put an empty diff in a block header extension? My understanding was that we prevented no-change proposals in block header extensions, but don't bother to try and avoid set/unset in subsequent blocks. So block header will always contain some kind of diff (or nothing). But a proposal of b1->P1v1, b2->P2v2, b3-P1v3 are all placed in block headers even though the end result is no change to the schedule but does result in a change to version.